Refactor checkout script

WalterKolczynski-NOAA opened the pull request:
**Description**
The checkout script is refactored. First, the repetitive checkout code is abstracted into a function that is called for each component. Second, if the clone already exists, checkout will still checkout the requested commit. Third, a new option (-c) is added that will delete any existing clones a create a fresh clone. These last two options allow updating of versions without needing to manually delete directories (and remembering which are created by checkout)

Additionally, a usage statement and code documentation are added.

There is no change to the existing command-line options, only new options. No change is necessary by users (although the checkout order and print messages have changed). The new syntax is as follows:
```
checkout.sh [-c][-h][-m ufs_hash][-o]
  -c:
      Create a fresh clone (delete existing directories)
  -h:
      Print this help message and exit
  -m ufs_hash:
      Check out this UFS hash instead of the default
  -o:
      Check out operational-only code (GTG and WAFS). Only authorized users can check out GTG.
```

Fixes #808

**Type of change**
- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**
- [x] Checkout on Hera as before
- [x] Checkout on Hera using -c to delete existing directories
- [x] Checkout on Hera to update version in-place
- [x] Checkout on Hera with non-existent hash to verify error-checking
  
**Checklist**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes

WalterKolczynski-NOAA committed:
Refactor checkout script

The checkout script is refactored. First, the repetitive checkout code
is abstracted into a function that is called for each component. Second,
if the clone already exists, checkout will still checkout the requested
commit. Third, a new option (-c) is added that will delete any existing
clones a create a fresh clone. These last two options allow updating of
versions without needing to manually delete directories (and remembering
which are created by checkout)

Additionally, a usage statement and code documentation are added.

Refs: #808

WalterKolczynski-NOAA committed:
Fix checkout locations

Checkout was failing to return to the top of the directory tree, so clones
wound up nested within each other.

Refs: #808

KateFriedman-NOAA commented:
Tested refactored checkout with `-o` option to check on GTG and WAFS checkout.

```
[Kate.Friedman@m71a1 sorc]$ sh checkout.sh -o
Received -o flag for optional checkout of operational-only codes
Starting checkout for ufs_model
|-- Cloning https://github.com/ufs-community/ufs-weather-model into ufs_model.fd
|-- Checking out Prototype-P8b of ufs_model
|-- Updating submodules of ufs_model (if any)

Starting checkout for gsi
|-- Cloning https://github.com/NOAA-EMC/GSI.git into gsi.fd
|-- Checking out a62dec6 of gsi
|-- Updating submodules of gsi (if any)

Starting checkout for gldas
|-- Cloning https://github.com/NOAA-EMC/GLDAS.git into gldas.fd
|-- Checking out gldas_gfsv16_release.v.1.28.0 of gldas
|-- Updating submodules of gldas (if any)

Starting checkout for ufs_utils
|-- Cloning https://github.com/ufs-community/UFS_UTILS.git into ufs_utils.fd
|-- Checking out 04ad17e of ufs_utils
|-- Updating submodules of ufs_utils (if any)

Starting checkout for verif-global
|-- Cloning https://github.com/NOAA-EMC/EMC_verif-global.git into verif-global.fd
|-- Checking out c267780 of verif-global
|-- Updating submodules of verif-global (if any)

Starting checkout for gfs_wafs
|-- Cloning https://github.com/NOAA-EMC/EMC_gfs_wafs.git into gfs_wafs.fd
|-- Checking out c2a29a6 of gfs_wafs
|-- Updating submodules of gfs_wafs (if any)

Checking out GTG extension for UPP

[Kate.Friedman@m71a1 sorc]$ ll logs/
total 19457
-rw-r--r-- 1 Kate.Friedman emcmodel 14857 May 25 17:43 build_ww3_prepost.log
-rw-r--r-- 1 Kate.Friedman emcmodel   555 May 25 17:41 checkout_gfs_wafs.log
-rw-r--r-- 1 Kate.Friedman emcmodel   549 May 25 17:41 checkout_gldas.log
-rw-r--r-- 1 Kate.Friedman emcmodel  1172 May 25 17:41 checkout_gsi.log
-rw-r--r-- 1 Kate.Friedman emcmodel   121 May 25 17:41 checkout_gtg.log
-rw-r--r-- 1 Kate.Friedman emcmodel  6763 May 25 17:38 checkout_ufs_model.log
-rw-r--r-- 1 Kate.Friedman emcmodel   540 May 25 17:41 checkout_ufs_utils.log
-rw-r--r-- 1 Kate.Friedman emcmodel   536 May 25 17:41 checkout_verif-global.log
```

The WAFS and GTG ops-only components checked out without issue:
```
[Kate.Friedman@m71a1 sorc]$ ll gfs_wafs.fd/
total 17
drwxr-sr-x  2 Kate.Friedman emcmodel 2048 May 25 17:41 driver
drwxr-sr-x  3 Kate.Friedman emcmodel  512 May 25 17:41 fix
drwxr-sr-x  2 Kate.Friedman emcmodel  512 May 25 17:41 jobs
drwxr-sr-x  3 Kate.Friedman emcmodel  512 May 25 17:41 modulefiles
drwxr-sr-x  3 Kate.Friedman emcmodel  512 May 25 17:41 parm
drwxr-sr-x  2 Kate.Friedman emcmodel 2048 May 25 17:41 scripts
drwxr-sr-x 11 Kate.Friedman emcmodel 2048 May 25 17:41 sorc
drwxr-sr-x  2 Kate.Friedman emcmodel  512 May 25 17:41 ush
[Kate.Friedman@m71a1 sorc]$ ll ufs_model.fd/FV3/upp/sorc/post_gtg.fd/
total 5632
-rw-r--r-- 1 Kate.Friedman emcmodel  24264 May 25 17:41 gtg_algo.F90
-rw-r--r-- 1 Kate.Friedman emcmodel 107457 May 25 17:41 gtg_compute.F90
-rw-r--r-- 1 Kate.Friedman emcmodel  32324 May 25 17:41 gtg_config.F90
-rw-r--r-- 1 Kate.Friedman emcmodel 134019 May 25 17:41 gtg.config.gfs
-rw-r--r-- 1 Kate.Friedman emcmodel  15573 May 25 17:41 gtg_ctlblk.F90
-rw-r--r-- 1 Kate.Friedman emcmodel  21034 May 25 17:41 gtg_filter.F90
-rw-r--r-- 1 Kate.Friedman emcmodel 931055 May 25 17:41 gtg_indices.F90
-rw-r--r-- 1 Kate.Friedman emcmodel   4831 May 25 17:41 gtg_smoothseams.F90
-rw-r--r-- 1 Kate.Friedman emcmodel  53096 May 25 17:41 map_routines.F90
-rw-r--r-- 1 Kate.Friedman emcmodel    542 May 25 17:41 README.md
```

Ran `build_all.sh` to check for build errors (none encountered):
```
[Kate.Friedman@m71a1 sorc]$ sh build_all.sh
Creating ../exec folder
 .... Building WW3 pre and post execs ....
 .... Building forecast model ....
 .... Building gsi ....
 .... Building UPP ....
 .... Building ufs_utils ....
 .... Building gldas ....
 .... Building gfs_wafs  ....
 .... Building workflow_utils ....
 .... Building gfs_util ....

 .... Build system finished ....
[Kate.Friedman@m71a1 sorc]$
```

WalterKolczynski-NOAA committed:
Tweak checkout stdout messages

Refs: #808

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/refactor_checkout

aerorahul reviewed (commented):
check GTG `manage_externals` being removed.
Other than that, looks cleaner.

WalterKolczynski-NOAA committed:
Fix typo in checkout

At some point, 'CLEAN' was mistakenly changed to just 'c' in the checkout
function, so existing directories were never deleted. This is now fixed.

Refs: #808

WalterKolczynski-NOAA committed:
Update checkout for GTG as submodule

The new UPP version used by UFS now uses a git submodule to include
the GTG code instead of manage_externals. Checkout script is now
updated to properly checkout the GTG code (for authorized users) when
the -o option is used.

Refs: #808

WalterKolczynski-NOAA commented:
@KateFriedman-NOAA The GTG procedure changed after I merged in the UFS model update (which included a UPP update). I've made the changes to update the checkout, if you would please confirm. It will *not* build, as there also need to be changes to the build (which I will handle separately).

I've requested GTG access so I can start testing myself, but I don't know how long that will take.
